source/grid/grid_tools.cc

```diff
@@ -1716,6 +1716,208 @@ namespace GridTools
+        subdomain_boundary_cell_enclosing_ball = (*subdomain_boundary_cell_iterator)->enclosing_ball();
+
+        subdomain_boundary_cells_centers.push_back( subdomain_boundary_cell_enclosing_ball.first);
+        subdomain_boundary_cells_radii.push_back( subdomain_boundary_cell_enclosing_ball.second);
+      }
+    AssertThrow( subdomain_boundary_cells_radii.size() == subdomain_boundary_cells_centers.size(),
+                 ExcInternalError());
+
+    // Find the cells that are within layer_thickness of predicate subdomain boundary
+    // distance but are inside the extended bounding box.
+    // Most cells might be outside the extended bounding box, so we could skip them.
+    // Those cells that are inside the extended bounding box but are not part of the
+    // predicate subdomain are possible candidates to be within the distance to the
+    // boundary cells of the predicate subdomain.
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      {
+        // Ignore all the cells that are in the predicate subdomain
+        if ( predicate(cell))
+          continue;
+
+        const std::pair<Point<spacedim>, double> &cell_enclosing_ball
+          = cell->enclosing_ball();
+
+        const Point<spacedim> &cell_enclosing_ball_center = cell_enclosing_ball.first;
+        const double &cell_enclosing_ball_radius = cell_enclosing_ball.second;
+
+        bool cell_inside = true; // reset for each cell
+
+        for (unsigned int d = 0; d < spacedim; ++d)
+          cell_inside &= (cell_enclosing_ball_center[d] + cell_enclosing_ball_radius > bounding_box.first[d])
+                         && (cell_enclosing_ball_center[d] - cell_enclosing_ball_radius < bounding_box.second[d]);
+        // cell_inside is true if its enclosing ball intersects the extended bounding box
+
+        // Ignore all the cells that are outside the extended bounding box
+        if (cell_inside)
+          for (unsigned int i =0; i< subdomain_boundary_cells_radii.size(); ++i)
+            if ( cell_enclosing_ball_center.distance_square(subdomain_boundary_cells_centers[i])
+                 <  Utilities::fixed_power<2>( cell_enclosing_ball_radius +
+                                               subdomain_boundary_cells_radii[i] +
+                                               layer_thickness + DOUBLE_EPSILON ))
+              {
+                active_cell_layer_within_distance.push_back(cell);
+                break; // Exit the loop checking all the remaining subdomain boundary cells
+              }
+
+      }
+    return active_cell_layer_within_distance;
+  }
+
+
+
+  template <class MeshType>
+  std::vector<typename MeshType::active_cell_iterator>
+  compute_ghost_cell_layer_within_distance ( const MeshType &mesh, const double layer_thickness)
+  {
+    std::function<bool (const typename MeshType::active_cell_iterator &)> predicate
+      = IteratorFilters::LocallyOwnedCell();
```

Comment:
I'm still nervous about the temporary here. You can avoid the issue by writing this as
```
  IteratorFilters::LocallyOwnedCell locally_owned_cell_predicate;
  std::function<bool (const typename MeshType::active_cell_iterator &)> predicate (locally_owned_cell_predicate);
```
This way the underlying object survives for longer than the function object.